### PR TITLE
Add globbing information to the profiler

### PR DESCRIPTION
**Pull request by smera**
Adds glob-specific profiling under /prof. Every top-level glob call will have an entry in the profiler report, plus a single entry with an aggregated item representing the overall evaluation time spent in glob calls.

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
I force pushed your branch for rebase to new master. I also added a commit to update the binary logger format.

File: ref/netstandard1.3/Microsoft.Build.Framework/Microsoft.Build.Framework.cs [hunk 1]
```diff
@@ -547,36 +547,45 @@ namespace Microsoft.Build.Framework.Profiler
...
+    }
+    public enum EvaluationLocationKind : byte
+    {
+        Condition = (byte)1,
+        Glob = (byte)2,
+        Item = (byte)0,
     }
     public enum EvaluationPass : byte
     {
-        InitialProperties = (byte)1,
```
Comment (AndyGerlicher) [inline ref/netstandard1.3/Microsoft.Build.Framework/Microsoft.Build.Framework.cs]: We should mark these explicitly so the value doesn't change when we add something new.

**Comment by AndyGerlicher** [comment]
If this gets in very soon I don't see a need to bump the file version. We need to do another build for CLI insertion I'd like to get this one in so version 5 might be around for a while (in preview form). FYI @KirillOsenkov

**Comment by AndyGerlicher** [comment]
@smera You put WIP... did you need to add anything else? Or can we merge this?

File: src/Framework/Profiler/EvaluationLocation.cs [hunk 2]
```diff
@@ -77,47 +93,56 @@ public struct EvaluationLocation
         public string ElementName { get; }
 
         /// <nodoc/>
-        public string ElementOrCondition { get; }
+        public string Description { get; }
```
Comment (cdmihai) [inline src/Framework/Profiler/EvaluationLocation.cs]: What's the difference between Description and EvaluationDescription?

**Comment by AndyGerlicher** [comment]
Merging this to take the change to the binary format. Please open a new PR to address the feedback. If we have to change the format again that's fine but we'd need to rev the version at that point. This is going in an official build for the CLI.

File: src/Framework/Profiler/EvaluationLocation.cs [hunk 3]
```diff
@@ -37,6 +39,19 @@ public enum EvaluationPass : byte
         Targets
     }
 
+    /// <summary>
+    /// The kind of the evaluated location being tracked
+    /// </summary>
+    public enum EvaluationLocationKind : byte
```
Comment (cdmihai) [inline src/Framework/Profiler/EvaluationLocation.cs]: There's more kinds than these 3. Like properties, imports, etc. Now, all of the others will get defaulted to item, which is incorrect.

Ways to fix:
- exhaustively describe kind (look at all the usings and generate a kind for each location)
- the above but using broader terms (e.g. if property evaluation has 2 usings, name them both as Property)
- Introduce an Unspecified kind for anything that's not these current 3, and in the future we'll specify on demand. Should not break binary compat.

**Comment by smera** [comment]
The idea was to add IDs and parent IDs to each entry, so in the future the binary logger can display things in a tree-like view.
This change doesn't need to get in now. It does change the structure of the message, so we'd need to bump the binary logger version 

---
In reply to: [351132774](https://github.com/Microsoft/msbuild/pull/2799#issuecomment-351132774) [](ancestors = 351132774)

File: src/Framework/Profiler/EvaluationLocation.cs (hunk 3 above)
Comment (smera) [reply src/Framework/Profiler/EvaluationLocation.cs]: Before this change we had just item and condition. I added glob, and that's why the enum.
I don't think we need more kinds at this point, so maybe this is just renaming 'item' to something else. Maybe 'element'?

---
In reply to: [156447130](https://github.com/Microsoft/msbuild/pull/2799#discussion_r156447130) [](ancestors = 156447130)

File: src/Framework/Profiler/EvaluationLocation.cs (hunk 2 above)
Comment (smera) [reply src/Framework/Profiler/EvaluationLocation.cs]: Renamed EvaluationDescription to EvaluationPassDescription, and Description to ElementDescription

---
In reply to: [156443221](https://github.com/Microsoft/msbuild/pull/2799#discussion_r156443221) [](ancestors = 156443221)

**Pull request merged**